File: 07-数据存储/05-图数据库/SchemaConstraints/selfcheck_schema_constraints.py
"""SchemaConstraints 自检：四类约束的建立、命名/幂等、写入校验。

官方来源（实读）：
  * Cypher Manual → Constraints（四类约束与 Edition 标注、graph type 提示）
  * Cypher Manual → Constraints → Syntax（CREATE CONSTRAINT 的完整语法：
    IS [NODE] UNIQUE / IS NOT NULL / IS :: <TYPE> / IS [NODE] KEY）
  * Cypher Manual → Constraints → Managing constraints（Key = 存在性+唯一性；
    LIST<STRING NOT NULL> 与 STRING | LIST<...> 联合类型；VECTOR<INT32>(42)；
    名字唯一；IF NOT EXISTS 的通知文案与三种命中情形）
  * Cypher Manual → Working with null（类型谓词对 null 恒为 true）
"""

from schema_constraints import (UNIQUENESS, EXISTENCE, TYPE, KEY, NODE, REL,
                                Database, ConstraintError, Entity,
                                _value_matches_type)

OK = 0
FAIL = []


def ck(name, cond, detail=""):
    global OK
    if cond:
        OK += 1
    else:
        FAIL.append("%s  %s" % (name, detail))


def eq(name, got, want):
    ck(name, got == want, "got=%r want=%r" % (got, want))


def raises(name, fn):
    try:
        fn()
    except ConstraintError:
        global OK
        OK += 1
        return
    FAIL.append("%s  未抛出 ConstraintError" % name)


def ok(name, fn):
    try:
        fn()
        global OK
        OK += 1
    except ConstraintError as e:
        FAIL.append("%s  意外报错: %s" % (name, e))


# ------------------------------------------------- 四类约束的建立
db = Database()
u = db.create_constraint("uniq_email", UNIQUENESS, NODE, "Person", ("email",))
eq("唯一性约束 kind", u.kind, UNIQUENESS)
e = db.create_constraint("ex_name", EXISTENCE, NODE, "Author", ("name",))
eq("存在性约束 kind", e.kind, EXISTENCE)
t = db.create_constraint("type_title", TYPE, NODE, "Movie", ("title",),
                         type_spec="STRING")
eq("类型约束 type_spec", t.type_spec, "STRING")
k = db.create_constraint("key_id", KEY, NODE, "Director", ("imdbId",))
eq("Key 约束 kind", k.kind, KEY)
r = db.create_constraint("uniq_order", UNIQUENESS, REL, "SEQUEL_OF", ("order",))
eq("关系约束 target", r.target, REL)

raises("TYPE 约束必须给类型",
       lambda: db.create_constraint("bad", TYPE, NODE, "Movie", ("title",)))
raises("EXISTENCE 只支持单属性",
       lambda: db.create_constraint("bad2", EXISTENCE, NODE, "A", ("x", "y")))
raises("唯一性至少一个属性",
       lambda: db.create_constraint("bad3", UNIQUENESS, NODE, "A", ()))

# ------------------------------------------------- 命名唯一（跨索引与约束）
raises("重名约束报错",
       lambda: db.create_constraint("uniq_email", KEY, NODE, "Person", ("x",)))
db.index_names.add("idx_taken")
raises("名字被索引占用也报错",
       lambda: db.create_constraint("idx_taken", EXISTENCE, NODE, "A", ("x",)))

# ------------------------------------------------- IF NOT EXISTS 三种情形
n0 = len(db.constraints)
same = db.create_constraint("uniq_email2", UNIQUENESS, NODE, "Person",
                            ("email",), if_not_exists=True)
eq("同模式不同名 → 不创建", len(db.constraints), n0)
eq("返回既有约束", same.name, "uniq_email")
# 官方 Example 29 的情形：与一个「不同类型的已有约束」同名 → 名字冲突优先
db.create_constraint("ex_name", UNIQUENESS, NODE, "Person", ("other",),
                     if_not_exists=True)
eq("同名不同类型 → 仍然不创建（名字冲突优先）",
   len(db.constraints), n0)
ck("该通知引用的是同名但不同类型的约束",
   any("uniq_email" in n and "ex_name" not in n for n in db.notifications)
   or any("ex_name" in n for n in db.notifications), str(db.notifications))
ck("产生了通知", len(db.notifications) >= 2, str(db.notifications))
ck("通知含 has no effect",
   any("has no effect" in n for n in db.notifications))

# 无 IF NOT EXISTS 时，同模式不同名会报错
raises("默认同模式不同名 → 报错",
       lambda: db.create_constraint("uniq_email9", UNIQUENESS, NODE, "Person",
                                    ("email",)))

# ------------------------------------------------- 存在性约束校验
d2 = Database()
d2.create_constraint("ex", EXISTENCE, NODE, "Author", ("name",))
ok("有属性 → 通过", lambda: d2.add(Entity(1, NODE, "Author", name="x")))
raises("缺属性 → 报错", lambda: d2.add(Entity(2, NODE, "Author")))
raises("属性值为 null → 报错",
       lambda: d2.add(Entity(3, NODE, "Author", name=None)))
ok("不匹配的标签不受约束", lambda: d2.add(Entity(4, NODE, "Book")))

# ------------------------------------------------- 类型约束校验
d3 = Database()
d3.create_constraint("tt", TYPE, NODE, "Movie", ("title",), type_spec="STRING")
ok("STRING 值通过", lambda: d3.add(Entity(1, NODE, "Movie", title="a")))
raises("INTEGER 值不通过",
       lambda: d3.add(Entity(2, NODE, "Movie", title=1)))
# 口径：值缺失/null 时类型谓词恒为 true（依据 Working with null）
ok("缺失属性 → 类型约束通过（null 的类型谓词恒 true）",
   lambda: d3.add(Entity(3, NODE, "Movie")))
ok("显式 null → 类型约束通过",
   lambda: d3.add(Entity(4, NODE, "Movie", title=None)))

# 复合类型：LIST<STRING NOT NULL> 与联合类型
ck("LIST<STRING NOT NULL> 接受字符串列表",
   _value_matches_type(["a", "b"], "LIST<STRING NOT NULL>"))
ck("LIST<STRING NOT NULL> 拒绝含 null 的列表",
   not _value_matches_type(["a", None], "LIST<STRING NOT NULL>"))
ck("LIST<STRING NOT NULL> 拒绝非列表",
   not _value_matches_type("a", "LIST<STRING NOT NULL>"))
ck("空列表满足 NOT NULL", _value_matches_type([], "LIST<STRING NOT NULL>"))
ck("联合类型 STRING|LIST<...> 接受字符串",
   _value_matches_type("a", "STRING | LIST<STRING NOT NULL>"))
ck("联合类型接受列表",
   _value_matches_type(["a"], "STRING | LIST<STRING NOT NULL>"))
ck("联合类型两者都不是则拒绝",
   not _value_matches_type(1, "STRING | LIST<STRING NOT NULL>"))
ck("BOOLEAN 不是 INTEGER", not _value_matches_type(True, "INTEGER"))
ck("INTEGER 值通过 INTEGER", _value_matches_type(1, "INTEGER"))
ck("VECTOR<INT32>(42) 接受序列", _value_matches_type([1] * 42,
                                                      "VECTOR<INT32>(42)"))

# ------------------------------------------------- 唯一性约束校验
d4 = Database()
d4.create_constraint("uq", UNIQUENESS, NODE, "Person", ("email",))
ok("首个写入通过",
   lambda: d4.add(Entity(1, NODE, "Person", email="a@b.c")))
raises("重复值 → 报错",
       lambda: d4.add(Entity(2, NODE, "Person", email="a@b.c")))
ok("不同值通过", lambda: d4.add(Entity(3, NODE, "Person", email="x@y.z")))
ok("不同标签不算冲突", lambda: d4.add(Entity(4, NODE, "Company",
                                             email="a@b.c")))
# 口径：缺属性不参与唯一性（官方本页未规定）
ok("缺属性不参与唯一性（口径）", lambda: d4.add(Entity(5, NODE, "Person")))
ok("第二个缺属性也不冲突（口径）", lambda: d4.add(Entity(6, NODE, "Person")))

# 复合唯一性
d5 = Database()
d5.create_constraint("uq2", UNIQUENESS, NODE, "P", ("a", "b"))
ok("复合首行通过", lambda: d5.add(Entity(1, NODE, "P", a=1, b=2)))
ok("复合仅一项相同不冲突", lambda: d5.add(Entity(2, NODE, "P", a=1, b=3)))
raises("复合全同才冲突", lambda: d5.add(Entity(3, NODE, "P", a=1, b=2)))

# ------------------------------------------------- Key = 存在性 + 唯一性
d6 = Database()
d6.create_constraint("k", KEY, NODE, "Director", ("imdbId",))
ok("Key 首个通过", lambda: d6.add(Entity(1, NODE, "Director", imdbId=1)))
raises("Key 拦重复", lambda: d6.add(Entity(2, NODE, "Director", imdbId=1)))
raises("Key 拦缺失（存在性部分）",
       lambda: d6.add(Entity(3, NODE, "Director")))
ok("Key 不拦不同值", lambda: d6.add(Entity(4, NODE, "Director", imdbId=2)))
# 与单纯唯一性约束的对比：唯一性不拦缺失，Key 会拦
d7 = Database()
d7.create_constraint("u", UNIQUENESS, NODE, "D", ("id",))
ok("纯唯一性不拦缺失", lambda: d7.add(Entity(1, NODE, "D")))

print("断言通过: %d" % OK)
if FAIL:
    print("失败 %d 条:" % len(FAIL))
    for f in FAIL:
        print("  - " + f)
    raise SystemExit(1)
print("ALL OK")

File: 07-数据存储/05-图数据库/SchemaConstraints/selfcheck_schema_constraints.go
// SchemaConstraints 自检：四类约束的建立、命名/幂等、写入校验。
package main

import (
	"fmt"
	"strings"
)

var okCount int
var failures []string

func ck(name string, cond bool, detail string) {
	if cond {
		okCount++
		return
	}
	failures = append(failures, fmt.Sprintf("%s  %s", name, detail))
}

func eqStr(name, got, want string) {
	ck(name, got == want, fmt.Sprintf("got=%s want=%s", got, want))
}

func eqInt(name string, got, want int) {
	ck(name, got == want, fmt.Sprintf("got=%d want=%d", got, want))
}

func mustErr(name string, err error) {
	if err != nil {
		okCount++
		return
	}
	failures = append(failures, name+"  未报错")
}

func mustOK(name string, err error) {
	if err == nil {
		okCount++
		return
	}
	failures = append(failures, fmt.Sprintf("%s  意外报错: %v", name, err))
}

func node(id int, entity string, props map[string]interface{}) *Entity {
	return &Entity{ID: id, Target: NodeT, Entity: entity, Props: props}
}

func main() {
	// ---------------------------------------------- 四类约束的建立
	db := NewDatabase()
	u, e1 := db.CreateConstraint("uniq_email", Uniqueness, NodeT, "Person",
		[]string{"email"}, "", false)
	mustOK("建唯一性约束", e1)
	eqStr("唯一性 kind", u.Kind, Uniqueness)
	_, e2 := db.CreateConstraint("ex_name", Existence, NodeT, "Author",
		[]string{"name"}, "", false)
	mustOK("建存在性约束", e2)
	t, e3 := db.CreateConstraint("type_title", TypeCon, NodeT, "Movie",
		[]string{"title"}, "STRING", false)
	mustOK("建类型约束", e3)
	eqStr("类型约束 typeSpec", t.TypeSpec, "STRING")
	k, e4 := db.CreateConstraint("key_id", KeyCon, NodeT, "Director",
		[]string{"imdbId"}, "", false)
	mustOK("建 Key 约束", e4)
	eqStr("Key kind", k.Kind, KeyCon)
	r, e5 := db.CreateConstraint("uniq_order", Uniqueness, RelT, "SEQUEL_OF",
		[]string{"order"}, "", false)
	mustOK("建关系约束", e5)
	eqStr("关系约束 target", r.Target, RelT)

	_, b1 := db.CreateConstraint("bad", TypeCon, NodeT, "Movie",
		[]string{"title"}, "", false)
	mustErr("TYPE 约束必须给类型", b1)
	_, b2 := db.CreateConstraint("bad2", Existence, NodeT, "A",
		[]string{"x", "y"}, "", false)
	mustErr("EXISTENCE 只支持单属性", b2)
	_, b3 := db.CreateConstraint("bad3", Uniqueness, NodeT, "A", nil, "", false)
	mustErr("唯一性至少一个属性", b3)

	// ---------------------------------------------- 命名唯一
	_, b4 := db.CreateConstraint("uniq_email", KeyCon, NodeT, "Person",
		[]string{"x"}, "", false)
	mustErr("重名约束报错", b4)
	db.IndexNames["idx_taken"] = true
	_, b5 := db.CreateConstraint("idx_taken", Existence, NodeT, "A",
		[]string{"x"}, "", false)
	mustErr("名字被索引占用也报错", b5)

	// ---------------------------------------------- IF NOT EXISTS
	n0 := len(db.Constraints)
	same, _ := db.CreateConstraint("uniq_email2", Uniqueness, NodeT, "Person",
		[]string{"email"}, "", true)
	eqInt("同模式不同名 → 不创建", len(db.Constraints), n0)
	eqStr("返回既有约束", same.Name, "uniq_email")
	// 官方 Example 29：与「不同类型的已有约束」同名 → 名字冲突优先
	db.CreateConstraint("ex_name", Uniqueness, NodeT, "Person",
		[]string{"other"}, "", true)
	eqInt("同名不同类型 → 仍然不创建", len(db.Constraints), n0)
	ck("产生了通知", len(db.Notifications) >= 2, fmt.Sprint(db.Notifications))
	ck("通知含 has no effect",
		len(db.Notifications) > 0 && strings.Contains(db.Notifications[0],
			"has no effect"), "")
	_, b6 := db.CreateConstraint("uniq_email9", Uniqueness, NodeT, "Person",
		[]string{"email"}, "", false)
	mustErr("默认同模式不同名 → 报错", b6)

	// ---------------------------------------------- 存在性
	d2 := NewDatabase()
	d2.CreateConstraint("ex", Existence, NodeT, "Author", []string{"name"}, "",
		false)
	mustOK("有属性通过", d2.Add(node(1, "Author", map[string]interface{}{
		"name": "x"})))
	mustErr("缺属性报错", d2.Add(node(2, "Author", map[string]interface{}{})))
	mustErr("值为 null 报错", d2.Add(node(3, "Author",
		map[string]interface{}{"name": nil})))
	mustOK("不匹配标签不受约束", d2.Add(node(4, "Book",
		map[string]interface{}{})))

	// ---------------------------------------------- 类型
	d3 := NewDatabase()
	d3.CreateConstraint("tt", TypeCon, NodeT, "Movie", []string{"title"},
		"STRING", false)
	mustOK("STRING 值通过", d3.Add(node(1, "Movie",
		map[string]interface{}{"title": "a"})))
	mustErr("INTEGER 值不通过", d3.Add(node(2, "Movie",
		map[string]interface{}{"title": 1})))
	mustOK("缺失属性通过（null 的类型谓词恒 true）", d3.Add(node(3, "Movie",
		map[string]interface{}{})))
	mustOK("显式 null 通过", d3.Add(node(4, "Movie",
		map[string]interface{}{"title": nil})))

	ck("LIST<STRING NOT NULL> 接受字符串列表",
		ValueMatchesType([]interface{}{"a", "b"}, "LIST<STRING NOT NULL>"), "")
	ck("LIST<STRING NOT NULL> 拒绝含 null",
		!ValueMatchesType([]interface{}{"a", nil}, "LIST<STRING NOT NULL>"), "")
	ck("LIST<STRING NOT NULL> 拒绝非列表",
		!ValueMatchesType("a", "LIST<STRING NOT NULL>"), "")
	ck("空列表满足 NOT NULL",
		ValueMatchesType([]interface{}{}, "LIST<STRING NOT NULL>"), "")
	ck("联合类型接受字符串",
		ValueMatchesType("a", "STRING | LIST<STRING NOT NULL>"), "")
	ck("联合类型接受列表",
		ValueMatchesType([]interface{}{"a"}, "STRING | LIST<STRING NOT NULL>"), "")
	ck("联合类型拒绝其它",
		!ValueMatchesType(1, "STRING | LIST<STRING NOT NULL>"), "")
	ck("BOOLEAN 不是 INTEGER", !ValueMatchesType(true, "INTEGER"), "")
	ck("INTEGER 值通过 INTEGER", ValueMatchesType(1, "INTEGER"), "")
	ck("VECTOR<INT32>(42) 接受序列",
		ValueMatchesType([]interface{}{1}, "VECTOR<INT32>(42)"), "")

	// ---------------------------------------------- 唯一性
	d4 := NewDatabase()
	d4.CreateConstraint("uq", Uniqueness, NodeT, "Person", []string{"email"},
		"", false)
	mustOK("首个写入通过", d4.Add(node(1, "Person",
		map[string]interface{}{"email": "a@b.c"})))
	mustErr("重复值报错", d4.Add(node(2, "Person",
		map[string]interface{}{"email": "a@b.c"})))
	mustOK("不同值通过", d4.Add(node(3, "Person",
		map[string]interface{}{"email": "x@y.z"})))
	mustOK("不同标签不算冲突", d4.Add(node(4, "Company",
		map[string]interface{}{"email": "a@b.c"})))
	mustOK("缺属性不参与唯一性（口径）", d4.Add(node(5, "Person",
		map[string]interface{}{})))
	mustOK("第二个缺属性也不冲突（口径）", d4.Add(node(6, "Person",
		map[string]interface{}{})))

	d5 := NewDatabase()
	d5.CreateConstraint("uq2", Uniqueness, NodeT, "P", []string{"a", "b"}, "",
		false)
	mustOK("复合首行通过", d5.Add(node(1, "P", map[string]interface{}{
		"a": 1, "b": 2})))
	mustOK("复合仅一项相同不冲突", d5.Add(node(2, "P",
		map[string]interface{}{"a": 1, "b": 3})))
	mustErr("复合全同才冲突", d5.Add(node(3, "P", map[string]interface{}{
		"a": 1, "b": 2})))

	// ---------------------------------------------- Key = 存在性 + 唯一性
	d6 := NewDatabase()
	d6.CreateConstraint("k", KeyCon, NodeT, "Director", []string{"imdbId"}, "",
		false)
	mustOK("Key 首个通过", d6.Add(node(1, "Director",
		map[string]interface{}{"imdbId": 1})))
	mustErr("Key 拦重复", d6.Add(node(2, "Director",
		map[string]interface{}{"imdbId": 1})))
	mustErr("Key 拦缺失", d6.Add(node(3, "Director",
		map[string]interface{}{})))
	mustOK("Key 不拦不同值", d6.Add(node(4, "Director",
		map[string]interface{}{"imdbId": 2})))
	d7 := NewDatabase()
	d7.CreateConstraint("u", Uniqueness, NodeT, "D", []string{"id"}, "", false)
	mustOK("纯唯一性不拦缺失", d7.Add(node(1, "D",
		map[string]interface{}{})))

	fmt.Printf("断言通过: %d\n", okCount)
	if len(failures) > 0 {
		fmt.Printf("失败 %d 条:\n", len(failures))
		for _, f := range failures {
			fmt.Println("  - " + f)
		}
		return
	}
	fmt.Println("ALL OK")
}
